### ISSUE TITLE
AssertionError crashes in PoolArena

### ISSUE THREAD

**Issue by pqrstuvwx** [netty/netty#12093]
Hello!

My team at AWS has a service that uses Netty 4.1. Ever since we upgraded from 4.1.48 to 4.1.52, we've been seeing occasional process (once every week or two) crashes due to AssertionErrors.

Here's the Netty portion of the stack trace:
```
  java.lang.AssertionError
  | at io.netty.buffer.PoolArena.tcacheAllocateSmall(PoolArena.java:165)
  | at io.netty.buffer.PoolArena.allocate(PoolArena.java:136)
  | at io.netty.buffer.PoolArena.allocate(PoolArena.java:128)
  | at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:378)
  | at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
  | at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
  | at io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:115)
```

We suspect that this is a bug introduced with [this commit](https://github.com/netty/netty/commit/7caced28c62f88f5045c3deaf8dfa4d4be1c803c), it's the only significant commit between 4.1.48 and 4.1.52 with relevant scope.

This is the failing assertion: https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PoolArena.java#L163

So the failure is either that the found subpage is marked for deletion, or the subpage's size is not as expected by sizeclasses. I'm leaning more towards the second possibility, since I think [this code](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PoolSubpage.java#L137) makes sure that the first subpage cannot be marked for deletion; but I may be missing something.

### Expected behavior
N/A
### Actual behavior
N/A
### Steps to reproduce
Unfortunately, we have not been able to find a repro; or any other indicators (metrics and logs) that predict the crash.

### Minimal yet complete reproducer code (or URL to code)
N/A

### Netty version
4.1.52

### JVM version (e.g. `java -version`)
openjdk version "11.0.13" 2021-10-19 LTS
OpenJDK Runtime Environment (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM (build 11.0.13+8-LTS, mixed mode)

### OS version (e.g. `uname -a`)
5.4.156-94.273.amzn2int.x86_64

**Comment by johnou** [comment]
@pqrstuvwx do you configure io.netty.allocator.directMemoryCacheAlignment? we ran into problems with the allocator as well which was fixed with by this commit.. https://github.com/netty/netty/pull/11987 (4.1.73.Final or above).

**Comment by pqrstuvwx** [comment]
Thanks - I did check #11987. We do not use `directMemoryCacheAlignment` (set to 0).

**Comment by normanmaurer** [comment]
@pqrstuvwx can you upgrade to latest netty version and paste the assert stack once it happens again ?

**Comment by johnou** [comment]
@normanmaurer if there's no performance penalty for using string concatenation in assert msgs could we print out the values so we know what failed and what state the arena was in?

**Comment by johnou** [comment]
> @normanmaurer if there's no performance penalty for using string concatenation in assert msgs could we print out the values so we know what failed and what state the arena was in?
> 
> ```java
> assert s.doNotDestroy && s.elemSize == sizeIdx2size(sizeIdx) : "doNotDestroy=" + s.doNotDestroy  + ", elemSize=" + s.elemSize + ", sizeIdx=" + sizeIdx;
> ```

for what it's worth I verified tostring is not invoked on objects in the assert debug message if they are enabled / disabled, until the assertion fails.

**Comment by johnou** [comment]
@pqrstuvwx were you able to confirm whether or not you can reproduce the assertion error in 4.1.74?

**Comment by pqrstuvwx** [comment]
Hi, 

Apologies for the late reply - we don't have any versions above 4.1.68 imported in our build system yet, so . Sorry to distract from the issue, but where would I find the source/javadocs jars for the latest versions? I need them to import the newer versions, bu they don't seem to be included with maven central releases anymore.

* https://repo1.maven.org/maven2/io/netty/netty-all/4.1.68.Final/ has source and javadocs jar
* https://repo1.maven.org/maven2/io/netty/netty-all/4.1.73.Final/ does not

I could try and build it myself, but if there's an official build I'd rather use that.

-Pete

**Comment by chrisvest** [comment]
@pqrstuvwx The `netty-all` module has become a pom-only module, and just pulls everything else in as transitive dependencies. So you have to download the modules you need individually.

**Comment by johnou** [comment]
@pqrstuvwx as you can see the modules with code in later versions are still shipped with sources and javadocs https://repo1.maven.org/maven2/io/netty/netty-buffer/4.1.74.Final/ (iirc it's actually a requirement in order to publish something to central Maven).

**Comment by johnou** [comment]
For a more robust approach I would recommend importing the netty-bom in dependencyManagement and then importing the required modules in your project modules.

**Comment by pqhwan** [comment]
thanks all, got the full context from https://github.com/netty/netty/pull/11732. Sorry - it took some time to work out the new packaging with the internal build system, didn't want to pull the rug out from under everyone :)

It'll take some time to build it in and push it out to testing stacks, but without a repro or an idea of the proximate cause, we'd just be waiting for weeks to see if the issue happens again. I'll keep this thread posted.

-Pete

**Comment by pqrstuvwx** [comment]
thanks all, got the full context from https://github.com/netty/netty/pull/11732. Sorry - it took some time to work out the new packaging with the internal build system, didn't want to pull the rug out from under everyone :)

It'll take some time to build it in and push it out to testing stacks, but without a repro or an idea of the proximate cause, we'd just be waiting for weeks to see if the issue happens again. I'll keep this thread posted.

**Comment by johnou** [comment]
@pqrstuvwx fwiw Netty 4.1.75.Final contains enhanced assertion error message to help diagnose your issue.

**Comment by pqhwan** [comment]
Thank you, we managed to get 4.1.73 imported and deployed on our end, so far it looks promising, no reoccurrence.

**Comment by pqhwan** [comment]
I'll update this thread once we've pushed the change to production, if we've still not seen any more of this issue then I think it can be closed out as having been fixed by 4.1.7x

**Comment by pqrstuvwx** [comment]
Thank you, we managed to get 4.1.73 imported and deployed on our end, so far it
  looks promising, no reoccurrence.
  I'll update this thread once we've pushed the change to production, if we've
  still not seen any more of this issue then I think it can be closed out as
  having been fixed by 4.1.7x

**Comment by trustin** [comment]
Sounds like it's safe to close now that it's not reproducible anymore since the upgrade, correct?
